### PR TITLE
More performance improvements

### PR DESCRIPTION
**Pull request by tlienart**
(I apologise for this big PR... but I think it's worth it, I would understand if it's not prioritized, the gist for working hard on this is that ST ends up being a fundamental op in MLJ and so, IMO, must be as optimized as possible)

This follows #77 and the discussion in #80. It is (mildly) **breaking** and so would require another minor release.

It brings **significant** performance improvements, the script ran is available in the `test/_benchmark.jl` script, condensed results below. In short, the performances are now on par with just calling `categorical` which is as good as we can make it.

```
@btime categorical($v1); # 28.495 ms (1000046 allocations: 34.34 MiB)
@btime coerce($v1, Multiclass);
#this branch ==> 28.755 ms (1000046 allocations: 34.34 MiB)
#master      ==> 46.380 ms (1000080 allocations: 31.48 MiB)

@btime categorical($v2); # 25.051 ms (799786 allocations: 28.22 MiB)
@btime coerce($v2, Union{Missing,Multiclass});
#this branch ==> 25.712 ms (799787 allocations: 28.22 MiB)
#master      ==> 139.152 ms (1599678 allocations: 51.69 MiB)

@btime coerce($df, autotype($df, :few_to_finite));
#this branch ==> 979.098 ms (11152394 allocations: 636.62 MiB)
#master      ==> 1.415 s (12273798 allocations: 609.21 MiB)
```

It doesn't affect `schema`. 

There are effectively two changes: a more careful treatment of missing values, leaving more responsibility to the user, this avoids unnecessarily checking whether a vector that has a Union{Missing,T} machine type really does have missing values. Second, a removal of compression which is explicitly discouraged in CategoricalArrays.

### Change  1 -- careful  treatment of missings

This is effectively  the result of the discussion at #80 . Basically it removes all checks for "real missing values". Two  important notes in changes of behaviour

- If the element machine type is `Union{Missing,T}`, the element scitype will be  `Union{Missing,S}` irrelevant of whether there are "true missings" or not
- An array of Any type coerced to some Finite type will have a `Union{Missing,<:Finite}` scitype irrelevant of whether it has missings or not

To clarify the second point, this is a behaviour that arises from `CategoricalArrays`: 

```julia
julia> a = categorical(Any[1,2,3]); eltype(a)
Union{Missing, CategoricalValue{Any,UInt32}}
```

### Change 2 -- no compression

The calls to `categorical` used to be like `categorical(v, true, ...)` where the  `true` corresponds to the `compress` argument. There may be a good reason to do  this but I don't immediately see it; the gain in memory is irrelevant especially as the previous behaviour required to have copies of the original vector by broadcasting over it. Also the use of that argument is [explicitly discouraged](https://juliadata.github.io/CategoricalArrays.jl/latest/using.html#CategoricalArrays.compress) due to performance issues resulting from type  instabilities. 

I may have overlooked some important reason why this is useful but in that case I suggest we reexport `CategoricalArrays.compress` and  let people who know what they're doing use it.

### Tightening

There may now be _more_ cases where a user would get scitype `Union{Missing,S}` where there are actually no missings and this could be a nuisance to users. I don't think this will arise very frequently but even infrequently I could see this causing confusion to a user. 
So I added a `tight` keyword to `coerce` (it can be passed to `autotype`) which would do what we were previously doing. Of course the use of it will go with a tax on performance.

```julia
julia> a = categorical(Any[1,2,3]); c = coerce(a, Multiclass); elscitype(c)
 Warning: Trying to coerce from `Any` to `Multiclass` with categoricals.
│ Coerced to Union{Missing,Multiclass} instead.
└ @ ScientificTypes ~/.julia/dev/ScientificTypes/src/conventions/mlj/utils.jl:4
Union{Missing, Multiclass{3}}

julia> a = categorical(Any[1,2,3]); c = coerce(a, Multiclass; tight=true); elscitype(c)
Multiclass{3}
```

### CODE REVIEW

**Comment by ablaom** [comment]
@tlienart Thanks for the work on this PR. I am sorry I cannot give a more postive review at this point.

According to the definition in the documents, which until now has been inviolate, *the scitype of an AbstractArray is AbstractArray{U} where U is the union of the element scitypes*. Of course, implementations do not literally need to compute element scitype unions, but this is the behaviour prescribed in the current design.

So if an array has no missing elements, then its scitype can *never* be `<:AbstractArray{Union{Missing,S}, N} where {S,N}`. As far as I am aware this definition is nowhere violated in the existing implementation. But in the proposed PR we already have (before considering categoricals or `Any` types):

```julia
scitype([1, 2, missing][1:2])
AbstractArray{Union{Missing, Count},1}
```

instead of 
```julia
julia> scitype([1, 2, missing][1:2])
AbstractArray{Count,1}
```

consistent with the definition.

A performance improvement is not enough to justify the proposed change. One would need to propose a new (presumably more complicated) definition and justify that first. (If categorical arrays had included `ordered` and `n_classes` as type parameters, then we could have defined scitypes as a simple map from ordinary types, rather than operating at the level of instances, but they didn't.)

If we accept the existing definition as inviolate, then to compute the scitype of any vector `v` for which `Missing <: eltype(v)` one cannot avoid actually parsing the array for missing values, as far as I can see.

**Comment by tlienart** [comment]
I understand your point but I don't think it's reasonable to ask the user to pay that price **all the times** when, in fact, the case would happen rarely.

I'd like to think more about this but here's an additional thought:

Let's say the user has a dataset X with (true) missings, then the user takes a subset of that dataset without true missings, trains a model on it. Then tries to predict on a chunk that has missings; the scitype would be different. I think this is an issue.

I would be in favour of adding an explicit keyword for `scitype` in the same way I added one for `coerce`. Also I feel the behaviour is restricted (`Union{Missing,T}` with no true missing and `Any`) and can be explained well. 

We can discuss this more but I think we really want to avoid ScientificTypes adding unnecessary overhead which users then assume is caused by "machine" and then generally assume that MLJ makes stuff slow. The type of performance gains indicated above should be considered in the light of "real-world" datasets where you would multiply this by the number of features which may be very large.

**Comment by ablaom** [comment]
What about the following alternative definition:

---

*The scitype of an `AbstractArray`, `A`, is always`AbstractArray{U}` where `U` is the union of the element scitypes,  with one exception: If `typeof(A) <: AbstractArray{Union{Missing,T}}` for some `T`, then the scitype is `AbstractArray{Union{Missing, U}}`, **even if `A` has no missing values**.* 

This exception is made for performance reasons. If one want's to override it, one uses `scitype(A, tight=true)`. 

---

Could we adhere strictly to this?

**Comment by tlienart** [comment]
Thanks for being flexible on this Anthony, your definition works fine with me but there is still the corner case of the `Any` type. 

For that one though, I think the story is less clear and I would be happy to  go  with the "broadcast by default" which would correspond to the behaviour that was there initially. For the sake of clarifying the discussion consider:

```julia
julia> using CategoricalArrays
julia> a = Any[1,2,3];
julia> ca = categorical(a)
3-element CategoricalArray{Any,1,UInt32}:
 1
 2
 3

julia> eltype(ca)
Union{Missing, CategoricalValue{Any,UInt32}}
```

What the current PR is doing:

1. AbstractArray{Any} => Continuous : broadcast `float` (warn if hasmissing that coercing to Union{Missing,Continuous})
1. AbstractArray{Any} => Count: broadcast `int` (warn if hasmissing that coercing to Union{Missing,Count})
1. AbstractArray{Any} => Multiclass or OrderedFactor : just call `categorical` (warn that coercing to Union{Missing, Multiclass} or Union{Missing, OrderedFactor})

Then

1. CategoricalArray{Any} => Count or Continuous: same story as with AbstractArray
1. CategoricalArray{Any} => Multiclass or OrderedFactor : just return the vector (possibly after ordering levels) with warning that coercing to Union{Missing,Multiclass} or Union{Missing,OrderedFactor}

Based on your input I'll adjust the current PR to take into account the new definition and have explicit tests for all those cases.

**Comment by ablaom** [comment]
Yes the coerce behaviour for `Any` sounds good. I think it's important to document it in the docstrings, if only to remind ourselves what that behaviour is. 

Please update the new  definition of scitype in /docs. You can just copy and paste what I wrote above.

**Comment by tlienart** [comment]
@ablaom, all done. Details of further changes below:

* docs adjustment based on your comment
* added a `tight` keyword to `scitype` and to  `schema` (on top of `coerce`)
* some code cleanup to not have to use the  `Val(Scitype(T, C))` trick, this makes calls simpler, the code more readable and reduces the need for specialised methods
* added a bunch of additional tests for the `tight` keyword as well as for the specific conversion cases with `Any` type discussed earlier
* updated Documenter to work with documenter 0.23 and 0.24

File: docs/src/index.md [hunk 1]
```diff
@@ -231,7 +238,7 @@ X = (x1=rand(10),
      x2=rand(10),
      x3=categorical(rand("abc", 10)),
      x4=categorical(rand("01", 10)))
-scitype(X)
+schema(X)
```
Comment (ablaom) [inline docs/src/index.md]: Sorry, why this change? We're demonstrating the definition of the `scitype` of a table, here.

Comment (ablaom) [reply docs/src/index.md]: I'm just going to make the minor reversion to previous state.
